vmdb2.1
.\" Copyright 2017 Lars Wirzenius <liw@liw.fi>
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.TH VMDB2 1
.SH NAME
vmdb2 \- create a disk image with Debian installed
.SH SYNOPSIS
.nh
.B vmdb2
.RB [ \-\-image\fR=\fIFILE ]
.RB [ \-\-output\fR=\fIFILE ]
.RB [ \-\-rootfs\-tarball\fR=\fIFILE ]
.RB [ \-h | \-\-help ]
.RB [ \-\-version ]
.RB [ \-v | \-\-verbose ]
.RB [ \-\-log\fR=\fIFILE ]
.RI [ FILE ]... ""
.hy
.B vmdb
.IR SPECFILE
.SH DESCRIPTION
.B vmdb2
reads a
.I "specification file"
that describes a disk image,
and builds the corresponding image.
The specification file uses YAML syntax,
and contains a list of steps.
Each step tells vmdb2 what to do next.
.SH OPTIONS
.TP
.BR \-h ", " \-\-help
show this help message and exit
.TP
.BR \-\-image =\fIFILE
use existing image file/device
.I FILE
.TP
.BR \-\-output =\fIFILE
create new image
.I FILE
(the size comes from the specification file)
.TP
.BR \-\-rootfs\-tarball =\fIFILE
store rootfs cache tar archives in
.I FILE
.TP
.BR \-v ", " \-\-verbose
verbose output
.TP
.BR \-\-log =\fIFILE
write log entries to
.I FILE
(default is to not write log files at all)
.TP
.BR \-\-version
show program's version number and exit
.SH EXAMPLE
To build a disk image:
.PP
.nf
.RS
vmdb2 foo.vmdb --output foo.img -v --rootfs-tarball foo.tar.gz
.RE
.fi
.SH "SEE ALSO"
.BR debootstrap (8).
.PP
.I https://vmdb2.liw.fi/
.PP
.IR /usr/share/doc/vmdb2/manual.html


NEWS
NEWS for vmdb2, the Debian disk image builder
=============================================================================

Version 0.19+git, not yet released
-----------------------------------------------------------------------------

* All use of the Python `cliapp` library has been dropped from vmdb2 now.


Version 0.19, released 2020-09-16
-----------------------------------------------------------------------------

Change in this release are entirely internal code changes that should
not affect users of vmdb2.

* Python Code has been formatted with the Black code formatting tool.

* Changes have been made towards dropping the dependency on the
  `cliapp` library: the Python standard library module `subprocess` is
  used instead of `cliapp.runcmd`, and the `cliapp` plugin system has
  been replaced with a custom one.


Version 0.18, released 2020-08-13
-----------------------------------------------------------------------------

* The way the manual was formatted could sometimes lead to different
  plugins' documentation being merged into the same section. This has
  now been fixed.


Version 0.17, released 2020-07-13
-----------------------------------------------------------------------------

* The luks plugin now runs the `cryptsetup` utility with the `-q`
  (alias for `--batch-mode`) option for batch mode, to prevent the
  utility from asking any questions of the user. Fix by Birger
  Schacht.


Version 0.16, released 2020-06-06
-----------------------------------------------------------------------------

* Progress output is now flushed after every write. This matters, when
  the output goes to a pipe, because then Python would buffer the
  output, meaning that progress output gets delayed until the buffer
  fills up, and that's useless.

* The documentation is now formatted using Subplot. Previously, the
  manual and the acceptance test suite were split into separate, but
  confusingly named files (vmdb2.md vs vmdb2.mdwn); now they are
  combined. The manual has been cleaned up a bit.

* Add a `quiet` field to the **grub step** to configure the kernel
  boot to be quiet or not. Default is now not quiet, which is a change
  from before.

* Add a `timeout` field to the **grub step** to configure grub menu
  timeout. Default is now zero seconds, which is a change from the
  Debian default of five seconds.

* Symbolic links can now be used in the **mkpart step** for device
  files, thanks to Tim Small.


Version 0.15, released 2020-05-05
-----------------------------------------------------------------------------

### New and changed features

* Document and rename the `virtual-filesystems` plugin. This allows
  installation of Debian packages that require `/proc` to be mounted
  to install, such as `ca-certificates-java`.
  
  Undo previous changes to have `/proc` automatically mounted in the
  chroot, when running commands in the chroot. That wasn't enough
  (e.g., it didn't cover Ansible), and `virtual-filesystems` seems
  generally like the more general and better solution.

  The `grub` step handles the filesystems it needs itself, but doesn't
  mind if they're already mounted.

  This change should NOT require changes to existing, working .vmdb
  files. If it does, please report a bug.

* The .vmdb file can now be passed in via the standard input, by
  specifying `-` as the filename. Patch from voney.

* The manual now documents that un-partitioned images are not
  supported, at least not yet.
  
* The "All went fine" message is now written at the very end of the
  vmdb2 run, if at all. Previously all the messages from teardown
  parts of steps were printed after it, hiding it effectively.
  
* The standard output and error output of the commands vmdb2 runs are
  no longer logged in real time, only after the command has finished.
  This is a side effect of the `cliapp.runcmd` helper function no
  longer being used. The Python standard library subprocess module is
  used directly instead. This change is a step to drop the dependency
  on the cliapp library.

### Bug fixes

* The `grub` step documentation now uses the correct key for
  specifying the image device (`image-dev`). Thanks: Sebastian
  Bachmann.

* The `mkpart` step now waits for the new device node to be created.
  Previously it sometimes took enough time to be created that the next
  step failed. This only affects when the output disk is a real block
  device, not a disk image file.

### Changes mainly affecting vmdb2 and plugin developers

* The `./check` script only runs the yarn integration testing tool if
  it's new enough to have the yarnutils Python library included. Yarn
  is being replaced with Subplot.

* Step runners must now use a new way of specifying expected and
  allowed keys for steps. The new way specifies types for mandatory
  keys, and default values for optional keys. The old
  `get_required_keys` method is gone. This change makes mistakes in
  step runner code a little less likely, and opens up the possibility
  of generating documentation about step keys automatically.

* Failures when un-mounting filesystems are now ignored. Such failures
  can happen for various reasons, but by far the most common reason is
  because the mount had already been unmounted and it's simpler to
  just ignore failures than keep meticulous track of what's been
  unmounted already.
  

Version 0.14.1
-----------------------------------------------------------------------------

This is a bug fix release to fix bugs in the 0.14 release that slipped
through due to insufficient testing.

* The shell plugin source file was misnamed. Fixed now.

* Fixes to problems found while building real images. The new
  `check-external` script builds my own images, and Debian's Raspberry
  Pi images.


Version 0.14, released 2020-03-21
-----------------------------------------------------------------------------

It has been two years since the previous release. These release notes
are a little sketchy.


vmdb2.py
import subprocess

def _runcmd(ctx, argv):
    p = subprocess.Popen(argv, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    stdout, stderr = p.communicate("")
    ctx['stdout'] = stdout
    ctx['stderr'] = stderr
    ctx['exit'] = p.returncode

def _binary(basename):
    return os.path.join(srcdir, basename)

def given_file(ctx, filename=None):
    with open(filename, 'wb') as f:
        f.write(get_file(filename))
 
def run_vmdb2(ctx, filename=None, output=None):
    vmdb2 = _binary('vmdb2')
    _runcmd(ctx, [vmdb2, filename, '-v', '--output', output])

def exit_code_is(ctx, exit_code=None):
    assert_eq(ctx['exit'], int(exit_code))

def stdout_contains(ctx, pat1=None, pat2=None):
    stdout = ctx.get('stdout', b'').decode('utf-8')
    i = stdout.find(pat1)
    assert i >= 0, "pat1 not found"
    i = stdout[i:].find(pat2)
    assert i >= 0, "pat2 not found after pat1"

def stdout_does_not_contain(ctx, pat1=None):
    stdout = ctx.get('stdout', b'').decode('utf-8')
    i = stdout.find(pat1)
    assert i == -1, "pattern found"
